### DISCUSSION TITLE
client.call_for return io_error

### DISCUSSION THREAD

**Discussion by jjqcat** [alibaba/yalantinglibs#641] [Q&A] [answered]
结构体参数传递和返回大小约100KB，独立线程ThreadFunc调用, 这个错误不是每次都出现，请帮忙分析下是否调用问题

async_simple::coro::Lazy<void> show_rpc_call(inPara, outPara)
{
  coro_rpc_client client;
[[maybe_unused]] auto ec = co_await client.connect("127.0.0.1", "8999");

 try
 {	
   auto resD = co_await client.call_for<coro_Work>(std::chrono::seconds(100), inPara);  //该函数执行时间可能要50秒以内

   if (resD.error().code != coro_rpc::errc::ok)
   {
     std::cout << std::endl << "*************************call error:" << resD.error().code << "*************************" << std::endl;

   }
}
catch (...)
 {
     client.close();
      std::cout << "error in call" << std::endl;
 }
}

}

ThreadFunc()
{

 syncAwait(show_rpc_call(inPara, outPara);

}

**Comment by poor-circle** [comment, marked as answer]
```cpp
if (resD.error().code != coro_rpc::errc::ok)
```
That is not correct. if rpc dont have error, call error() method is undefiend behavior(). See https://en.cppreference.com/w/cpp/utility/expected/error

**Comment by poor-circle** [comment]
You can check expected<T,E> by:
```cpp
if (resD/*or: resD.has_value()*/)
{
  resD.value();
   //it's OK
}
else {
  resD.error();
  // error happen
}
```
OR:
```cpp
std::cout<<resD.value()<<std::endl; //throw expection if error happened.
```

**Comment by poor-circle** [comment]
你可以验证一下到底是真的有rpc错误，还是你在没有错误发生的时候去调用了error()方法。

**Comment by jjqcat** [comment]
非常感谢，加上if (resD）条件后，没有失败和错误了。再请教一个问题，这个通讯安全性需要考虑吗，如
客户端通过RPC调用远程函数，是否有外部非法客户端（模拟合法客户端）【通过反编译破解技术，确认了远程IP和端口】，调用远程RPC。 或者有什么好的建议

**Comment by poor-circle** [comment]
> 非常感谢，加上if (resD）条件后，没有失败和错误了。再请教一个问题，这个通讯安全性需要考虑吗，如 客户端通过RPC调用远程函数，是否有外部非法客户端（模拟合法客户端）【通过反编译破解技术，确认了远程IP和端口】，调用远程RPC。 或者有什么好的建议

这个就像http一样，你需要自己实现一些权限校验机制。比如要求用户在连接时先登陆，然后其他rpc函数检测当前是否有权限。
另外，coro_rpc支持使用ssl/tls加密。

**Comment by poor-circle** [comment]
> 非常感谢，加上if (resD）条件后，没有失败和错误了。再请教一个问题，这个通讯安全性需要考虑吗，如 客户端通过RPC调用远程函数，是否有外部非法客户端（模拟合法客户端）【通过反编译破解技术，确认了远程IP和端口】，调用远程RPC。 或者有什么好的建议

此外，请注意，由于客户端处于不可信环境下，因此客户端的一切计算结果同样是不可信的，这点和http或其他rpc服务没有区别，服务端都应该对客户端的计算结果做校验。

**Comment by jjqcat** [comment]
谢谢！ 这个服务端有办法获取客户端的IP等信息吗
